### DISCUSSION TITLE
PEP520 Preserving Class Attribute Definition Order using 2.0 ORM

### DISCUSSION THREAD

**Discussion by CodeBardian** [sqlalchemy/sqlalchemy#11913] [Usage Questions]
I am in the process of migrating an existing ORM mapping to the new 2.0 style.
My application is relying on the PEP520 promise of preserving class attibute definition order.

> PEP 520: Preserving Class Attribute Definition Order
Attributes in a class definition body have a natural ordering: the same order in which the names appear in the source. This order is now preserved in the new class’s __dict__ attribute.

This condition was previously met when using a 1.4 mapping. Now in 2.0 when mixing annotated and non-annotated columns this does not hold anymore. Neither when using `Model.__dict__` nor in `Mapper.all_orm_descriptors`.

Minimal example below:

```
from sqlalchemy import String, Integer
from sqlalchemy.orm import DeclarativeBase, mapped_column, Mapped

class Base(DeclarativeBase):
    pass

class A(Base):
    __tablename__ = 'a'
    
    id = mapped_column(Integer, primary_key=True)
    name = mapped_column(String)
    description = mapped_column(String(10))
    

class B(Base):
    __tablename__ = 'b'
    
    id = mapped_column(Integer, primary_key=True)
    name: Mapped[str]
    description: Mapped[str] = mapped_column(String(10))
    
    
for prop in A.__mapper__.all_orm_descriptors:
    print(prop)
    
for prop in B.__mapper__.all_orm_descriptors:
    print(prop)

```

this prints
```
A.id
A.name
A.description

B.id
B.description  # !!!
B.name
```

To make use of type hinting, I am now finding myself replacing all simple instances of `col = Column(String)` with `col: Mapped[Optional[str] = mapped_column()` which seems way more convoluted. 

Is the PEP520 something you have already considered or would consider when mixing in the annotated style?

I see that there are a couple of tests for pep520 but only using non-annotated columns.
https://github.com/sqlalchemy/sqlalchemy/blob/64c1299180c2d944142d54bea741355d474bcbde/test/orm/test_inspect.py#L516

**Comment by CaselIT** [comment]
Hi,

I think the issue is that `name: Mapped[str]` does not generate an entry in the class dict while the class is defined, it's added later by the init_subclass of DeclarativeBase

As far as I know there isn't a way of getting the order of the attributes after the fact, so I'm not sure how it could be solved.

Can you confirm that using `name: Mapped[str] = mapped_column()` does create things in the proper order?

**Reply by CodeBardian** [reply]
I can confirm that using `name: Mapped[str] = mapped_column()` does generate the desired order. Even `name: Mapped[str] = None` seems to work, i just realized. So that will probably be my option going forward. 

Just as an additional note when using [SQLModel](https://github.com/fastapi/sqlmodel) which also uses type annotations for column declaration the order seems to be preserved. I didn't investigate further but I guess it proves that there has to be a way to get the correct column order even when mixing the column declaration using the annotated style.

```
from typing import Optional

from sqlmodel import Field, SQLModel
from sqlalchemy.orm import DeclarativeBase


class C(SQLModel, table=True):
    id: int = Field(default=None, primary_key=True)
    name: str
    description: str = Field(max_length=10)
    
    
for prop in C.__mapper__.all_orm_descriptors:
    print(prop)

```

this prints the columns in the desired pep520 order
```
C.id
C.name
C.description
```

**Reply by CaselIT** [reply]
> Even name: Mapped[str] = None seems to work, i just realized. So that will probably be my option going forward.

Please don't, that's not supported and there is no guarantee it will keep working. Also the type checker will not be happy and that's the main reason for using the new style.

If you want something shorter alias `mapped_column` to `mc`

> Just as an additional note when using [SQLModel](https://github.com/fastapi/sqlmodel) ...

As far as I know it requires types in all attributes, your example does not use types in every attribute. If that's expected you could use `__annotations__` for the order. SQLAlchemy can't since it needs to support cases like `id = mapped_column(Integer, primary_key=True)` that will not be in the annotations dict

**Reply by CodeBardian** [reply]
alright, thanks for the insight and quick reply
